Documentation: SHAKE (#75)
sehlen-bsi/botan-docs: cryptodoc/src/01_hash.rst

.. [#sha_intrinsics]
   https://github.com/noloader/SHA-Intrinsics

SHAKE
-----

> Botan implements the two SHA-3 XOFs SHAKE128 and SHAKE256 in
> :srcref:`src/lib/hash/shake/shake.cpp` as defined in [FIPS-202]_. It
> utilizes the sponge construction methods of the basic SHA-3
> implementation using the SHAKE-specific padding. In contrast to SHA-3
> we allow arbitrary output lengths which are provided to the constructor
> of the class.

Reviewer: I think this is basically the same discussion we were having with Falko and Jack: Technically, both SHA-3 and SHAKE rely on the Keccak construction as defined by FIPS with slightly different parameterizations.

I don't have a strong opinion on this. Nevertheless, we could already choose the wording accordingly, as Falko's work (once merged) will also reflect that in the code structure.

On the other hand, your wording stays closer to the implementation in Botan 3.1.1. It reflects that the SHAKE code uses the static methods `SHA_3::permute()` and friends.

```suggestion
Botan implements the two SHA-3 XOFs SHAKE128 and SHAKE256 in
:srcref:`src/lib/hash/shake/shake.cpp` as defined in [FIPS-202]_. It
utilizes the Keccak sponge construction methods also used in the SHA-3
implementation using a different padding that is specific to SHAKE. In contrast to SHA-3
it allows arbitrary output lengths which are provided to the constructor
of the class.
```

Author: I like this more generic version. It should also be valid for the KMAC PR version. Applied in 768d44a.
